### PR TITLE
Track the reason for object reconciliation

### PR DESCRIPTION
**Pull request by teozkr**
See https://github.com/clux/kube-rs/issues/457#issuecomment-869181430, #423 

Generally this only tracks the *first* reason an object to be reconciled.

### CODE REVIEW

File: kube-runtime/src/controller/mod.rs [hunk 1]
```diff
@@ -136,6 +155,61 @@ impl<T> Context<T> {
...
+    Clone(bound = "K::DynamicType: Clone"),
+    PartialEq(bound = "K::DynamicType: PartialEq"),
+    Eq(bound = "K::DynamicType: Eq"),
+    Hash(bound = "K::DynamicType: Hash")
+)]
+pub struct ReconcileRequest<K: Resource> {
+    pub obj_ref: ObjectRef<K>,
+    #[derivative(PartialEq = "ignore", Hash = "ignore")]
+    pub reason: ReconcileReason,
+}
```
Comment (clux) [inline kube-runtime/src/controller/mod.rs:175]: ah the core idea. i like it. smart way of injecting data without adding more to the applier.

File: kube-runtime/src/controller/mod.rs [hunk 2]
```diff
@@ -136,6 +155,61 @@ impl<T> Context<T> {
...
+
+#[derive(Debug, Clone)]
+pub enum ReconcileReason {
+    Unknown,
+    ObjectUpdated,
+    RelatedObjectUpdated { obj_ref: ObjectRef<DynamicObject> },
+    ReconcilerRequestedRetry,
+    ErrorPolicyRequestedRetry,
+    BulkReconcile,
+    Custom { reason: String },
```
Comment (clux) [inline kube-runtime/src/controller/mod.rs:194]: reasons look sensible. but curious about the `Custom` one, it's not created anywhere, but presumably intentioned as user-facing? can they be returned from the reconciler?

**Review by clux — APPROVED** [review]
this is a really clean way of doing it. looking forward to seeing how the spans look!

left a minor nit on clippy, but can fix that up later, feel free to merge.

**Comment by clux** [comment]
this probably closes #457 with all other tasks done.

File: kube-runtime/src/controller/mod.rs [hunk 3]
```diff
@@ -363,12 +446,17 @@ where
...
     where
-        Child::DynamicType: Debug + Eq + Hash,
+        Child::DynamicType: Debug + Eq + Hash + Clone,
     {
-        let child_watcher = trigger_owners(try_flatten_touched(watcher(api, lp)), self.dyntype.clone());
+        let child_watcher = trigger_owners(
+            try_flatten_touched(watcher(api, lp)),
+            self.dyntype.clone(),
+            dyntype,
+        );
```
Comment (clux) [inline kube-runtime/src/controller/mod.rs:459]: actually question: why does `dyntype` need to be explicitly passed in now?

Comment (clux) [reply kube-runtime/src/controller/mod.rs:459]: nevermind, i see it's for getting info for the reason. damnit, i thought we had that info before. that's kind of annoying. i'm slightly sad that i moved dyntype out of `Api` now.

**Review by kazk — APPROVED** [review]

File: kube-runtime/src/controller/mod.rs (hunk 2 above)
Comment (teozkr) [reply kube-runtime/src/controller/mod.rs:194]: Yes, if you're using `applier` then custom trigger functions can inject custom reasons.

File: examples/crd_derive.rs [hunk 4]
```diff
@@ -37,6 +37,7 @@ pub struct FooStatus {
 
 fn main() {
     println!("Kind {}", FooCrd::kind(&()));
+    #[allow(clippy::blacklisted_name)]
```
[the remark this answers is not in the document]
Comment (teozkr) [reply examples/crd_derive.rs:40]: The annoying part there is that it only applies to CI builds. I have Rust-Analyzer set up to watch-run clippy, which wouldn't pick that up.

We could do it with a crate-global `#![allow(clippy::blacklisted_name)]`, but I'm not sure that would be a huge win (it'd still have to be applied separately to each example, kind of going against the whole point).

**Comment by clux** [comment]
released in 0.58 :-)

**Pull request merged**